apiserver/common/charms.go

```diff
@@ -109,3 +111,21 @@ func ValidateCharmOrigin(o *params.CharmOrigin) error {
 	}
 	return nil
 }
+
+// CharmLocatorFromURL returns a CharmLocator using the charm name, revision
+// and source (which is extracted from the schema) of the provided URL.
+func CharmLocatorFromURL(url string) (charm.CharmLocator, error) {
+	u, err := internalcharm.ParseURL(url)
+	if err != nil {
+		return charm.CharmLocator{}, errors.Trace(err)
+	}
+	source, err := charm.ParseCharmSchema(internalcharm.Schema(u.Schema))
+	if err != nil {
+		return charm.CharmLocator{}, errors.Trace(err)
+	}
+	return charm.CharmLocator{
+		Name:     u.Name,
+		Revision: u.Revision,
+		Source:   source,
+	}, nil
+}
```

Comment:
Move this to `apiserver/internal/charms/charms.go`

Reply:
Done.

Reply:
Don't know what happened with my commit. Moved now.